fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -427,6 +261,178 @@ def _threshold_optimization_equalized_odds(sensitive_features, labels, scores, g
     return predicted_EO_by_sensitive_feature
 
 
+_SUPPORTED_MITIGATIONS = {
+    DEMOGRAPHIC_PARITY: _threshold_optimization_demographic_parity,
+    EQUALIZED_ODDS: _threshold_optimization_equalized_odds}
+
+
+def _get_soft_predictions(estimator, X):
```

Comment:
right now we're using `predict`, not `predict_proba` for this. However, when the scores are binary this method doesn't work as well as when using real-value scores so for examples (such as one of the notebooks) we've actually done this manually by wrapping the estimator and mapping `predict_proba` to `predict`. I'm not familiar with `decision_function`, but would we just use `predict` if the others aren't available?

Reply:
Also adding @MiroDudik  for awareness

Reply:
Interesting. As @romanlutz says, previously we were only using `predict` as the first stage of the pipeline before applying thresholding. I think it's reasonable to consider the back-off sequence `predict_proba` -> `decision_function` -> `predict`, where in `predict_proba` we only look at the predicted probability of the positive class.

So maybe two fixes to the below:
* for `predict_proba`, just extract the second column
* after `else`, don't throw an error, but instead return `predict`

Reply:
`predict` has very different semantics than `predict_proba` or `decision_function`. The probabilities `predict_proba` gives usually come from the raw values that the `decision_function` gives. You can think of the `decision_function` values as the distances from the hyperplanes in the case of an SVM.

`predict` always gives categorical results, and are not smooth. Either the method (`ThresholdOptimizer`) knows how to deal with that or it doesn't. If it doesn't, I don't think falling back to `predict` is a good idea. The `predict` function of a classifier shouldn't be giving soft outputs.

Alternatively, we could ask the user to either pass a classifier with a `predict_proba` or a `decision_function` provided, or a regression on a binary classifier which gives soft outputs as the predictions. One of our classifiers uses a regressor under the hoods in sklearn.

Reply:
I'm lacking relevant sklearn knowledge here: does every classifier have either `predict_proba` or `decision_function` or are there any that don't?
I don't understand what you mean by regression on a binary classifier. How would that be called? In some examples we use `predict_proba` from `LogisticRegression` and take the second entry (assuming binary classification) which is the probability for label 1.

Reply:
`ThresholdOptimizer` has no problem handling hard 0/1 predictions and that's in fact what the original paper was assuming. That's also the reason why we used `predict` in our original implementation. However, in those cases when the classifier is also providing a soft prediction (in the form of `decision_function` or `predict_proba`), it is obviously better to use that.

So I definitely think that we should allow users to run `ThresholdOptimizer` based on `predict` of the underlying classifier. What I wonder about is whether we should allow folks to force `ThresholdOptimizer` to use `predict` even in those cases when the classifier also implements the soft versions.

PROS: it allows you to run the "textbook version" of the algorithm
CONS: it's essentially guaranteed to lead to inferior performance

Reply:
> I don't understand what you mean by regression on a binary classifier.

one example is that the `GradientBoostingClassifier` uses `DecisionTreeRegressor` at each step instead of a `DecisionTreeClassifier`. In some cases people use a regressor on the binary classification problem with 0/1 as the target, and they use the soft predicted output as a proxy for _probability_. 

We can have a `predict_method` parameter for the users to set it to whatever they want, and have it as `auto` by default.

Reply:
I like the final suggestion! Let's go with `predict_method` parameter defaulted to `auto`. For `auto`, we can try `predict_proba` -> `decision_function` -> `predict` (also I don't mind dropping the final `predict`).

The only weirdness is that whenever `predict_method` is `predict_proba`, we will need to extract the second column of the output.